### PR TITLE
Delegate submit events until amp-form is loaded 

### LINKED ISSUE #6916: amp-form on=form.submit won't work for GET non-XHR requests
The call we make to `this.handleSubmit_` in `amp-form` doesn't actually invoke `form.submit()` and since this isn't an actual submission event we'd need to trigger it manually in that case.

```javascript
if (!e && GET && non-XHR) {
  // after doing validation checks.
  this.form_.submit();
}
```

### LINKED ISSUE #6928: XHR submission should be delayed until amp-form is loaded
To make sure submissions made while `amp-form` is still being loaded go through, all XHR submissions will be delegated through AMP's own action service and executed once `amp-form` install the action handler.

This will be followed up with a fix for #6927

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #6928 
Fixes #6916 

This also fixes non-xhr submit throguh the `on=action:form.submit` syntax.

### CODE REVIEW

File: build-system/server.js [hunk 1]
```diff
@@ -487,6 +487,20 @@ app.use(['/examples/*', '/extensions/*'], function (req, res, next) {
   next();
 });
 
+/**
+ * Append ?sleep=5 to any included JS file in examples to emulate delay in loading that
```
Comment (cramforce) [inline build-system/server.js]: This is super awesome!!!!!

Comment (camelburrito) [inline build-system/server.js]: +1

File: src/document-submit.js [hunk 2]
```diff
@@ -111,4 +112,18 @@ export function onDocumentFormSubmit_(e) {
   if (actionXhr) {
     checkCorsUrl(actionXhr);
   }
+
+  // For xhr submissions relay the submission event through action service to
+  // allow us to wait for amp-form (and possibly its dependencies) to execute
+  // the actual submission. For non-XHR GET we let the submission go through
+  // to allow _blank target to work.
+  if (actionXhr) {
+    actionServiceForDoc(form).execute(form, 'submit', null, form, e);
```
Comment (camelburrito) [inline src/document-submit.js]: i think this should be the last line in this block

File: extensions/amp-form/0.1/amp-form.js [hunk 3]
```diff
@@ -276,6 +278,11 @@ export class AmpForm {
       for (let i = 0; i < varSubsFields.length; i++) {
         this.urlReplacement_.expandInputValueSync(varSubsFields[i]);
       }
+
+      // If this wasn't a submission event, trigger form submit.
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I don't really understand this. Could you explain? Can we split this into two methods?

File: src/document-submit.js (hunk 2 above)
Comment (dvoytenko) [inline src/document-submit.js]: Annotate `null`

File: extensions/amp-form/0.1/amp-form.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Yeah this confusion comes from reusing this method once for actual `submit` event handling and the other for doing an actual submission flow (triggered by `on=form.submit` for example).

I couldn't figure out how to refactor the function to reuse the functionality in either cases, but I can think a bit more about it and see if I can simplify it.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Ok, I split this method to multiple ones for clarity and handling a submission event is now separate from handling a submission action. PTAL 👀

File: src/document-submit.js (hunk 2 above)
Comment (mkhatib) [inline src/document-submit.js]: Done

Comment (mkhatib) [inline src/document-submit.js]: Done

**Comment by mkhatib** [comment]
@dvoytenko PTAL 👀

File: extensions/amp-form/0.1/amp-form.js [hunk 4]
```diff
@@ -161,13 +163,14 @@ export class AmpForm {
...
-      this.handleSubmit_();
+      this.handleSubmitAction_();
     }
   }
 
   /** @private */
   installEventHandlers_() {
-    this.form_.addEventListener('submit', e => this.handleSubmit_(e), true);
+    this.form_.addEventListener(
+        'submit', e => this.handleSubmitEvent_(e), true);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: `this.handleSubmitEvent_.bind(this)`

File: extensions/amp-form/0.1/amp-form.js [hunk 5]
```diff
@@ -179,106 +182,149 @@ export class AmpForm {
...
    * browser submit event blocking. Otherwise any other submit listeners would get the
    * event.
    *
    * For example, action service shouldn't trigger 'submit' event if form is actually
-   * invalid. stopImmediatePropagation allows us to make sure we don't trigger it
+   * invalid. stopImmediatePropagation allows us to make sure we don't trigger it.
    *
-   *
-   * @param {?Event=} opt_event
+   * @param {Event} event
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: `!Event`

File: extensions/amp-form/0.1/amp-form.js [hunk 6]
```diff
@@ -179,106 +182,149 @@ export class AmpForm {
...
-        opt_event.preventDefault();
-      }
+  handleSubmitEvent_(event) {
+    if (this.state_ == FormState_.SUBMITTING || !this.checkValidity_()) {
+      event.stopImmediatePropagation();
+      event.preventDefault();
       return;
     }
+    if (this.xhrAction_ || this.method_ == 'POST') {
+      event.preventDefault();
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Please document why `preventDefault` is not needed in other situations.

**Comment by dvoytenko** [comment]
@mkhatib Looks good. Just a couple of minor requests.

File: extensions/amp-form/0.1/amp-form.js (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

**Comment by mkhatib** [comment]
PTAL 👀

**Pull request merged**